### ISSUE TITLE
[Feature Request] Community Requested Payment Gateways

### ISSUE THREAD

**Issue by BelleNottelling** [FOSSBilling/FOSSBilling#1360]
In an attempt to help organize the existing requests / issues open on GitHub, I've decided we should merge all payment gateway requests into one issue. Additionally, this can help serve as a place where we can evaluate the popularity and usefulness of a specific request to determine if it's something we may want to consider official support for.

## Community Requested Payment Gateways:
 - [Razorpay](https://razorpay.com/)
     - Supports Indian markets. Seems to be largely popular with a valuation of $7.5 billion in 2021. According to [ycombinator](https://www.ycombinator.com/companies/razorpay) "Razorpay enables payments and banking solutions for over 10Mn businesses"
     - Of the existing Payment Gateways, PayPal, Stripe, and 2Checkout supports India. I do not know if the 2Checkout payment gateway works
 -  [Paystack](https://paystack.com/)
    - Support African markets. I'm unsure about their popularity, however Stripe did purchase them in 2020 so they must be fairly popular.
    - Paystack themselves have a [payment gateway](https://github.com/PaystackHQ/plugin-boxbilling) that was built for BoxBilling, however it doesn't seem to be maintained and will no longer work with FOSSBilling.
    - There is also a community made [payment gateway](https://github.com/samapraku/boxbilling-paystack-adapter) for Paystack. It's been updated recently, however it is currently incompatible with FOSSBilling version 0.5.0
 -  [Adyen](https://www.adyen.com/)
    - Supports quite a lot of countries, several larger companies are their customers (a few listed on the website)
    -  No existing payment gateway.
 - [Cardinity](https://cardinity.com/)
    - Seems to primarily focus on supporting European markets.
    - Apparently used to have a payment gateway for BoxBilling, however it no longer exists.
    - They seem to maintain a fair few payment gateways themseves, so it's possible they'd consider making one for FOSSBilling once we are production-ready.
 -  [Mercado Pago](https://www.mercadopago.com.ar/)
    - Supports Latin America
    - Is popular there from my understanding.
    - No existing integrations I could find.
 - [Flutterwave](https://flutterwave.com/us/)
   - Looks to provide support for countries our existing Payment Gateways are lacking for. (Nigeria for example)
   - Decently popular, processes an average of 500k transactions per day (according to their own website)
   - No existing integration.
 - [BTCPay](https://btcpayserver.org/)
   - Cryptocurrency payment gateway and as such should support anyone anywhere in the world.
   - Looks to be pretty popular with 5k stars on GitHub.
   - ❤️ Open source!
   - No existing payment gateway.
 - [Efí (previously called Gerencianet) ](https://sejaefi.com.br/)
   - Targets Brazil
   - Unsure about the popularity of it
   - They do have an old [payment gateway](https://github.com/gerencianet/gn-api-boxbilling) they created for BoxBilling, however a FOSSBilling user reported it no longer works. We may want to consider contacting them and asking they update the existing one for FOSSBilling.

### Notes
 - We could consider using something such as [Omnipay](https://github.com/thephpleague/omnipay), however it appears that Omnipay development has more or less stopped and most things haven't been updated in years.
 - I don't know how practically we can build payment gateways for anything that doesn't provide support for countries we live in. Our ability to do so will depend on what languages their documentation is available in and if we could even create an account for.

**Comment by tobsowo** [comment]
I'll love to see Paystack and Flutterwave in that order basically works for most African countries.

**Comment by BelleNottelling** [comment]
> I'll love to see Paystack and Flutterwave in that order basically works for most African countries.

Thanks for the info. That definitely helps us understand which ones are more valuable to a larger group of people compared to other options

**Comment by John-S4** [comment]
Nice job compiling the list @BelleNottelling - I think it's a lot more useful to have them all here than to have multiple feature requests scattered around. 

A few thoughts...

It might be useful to also add a list of the currently supported payment gateways and if they are maintained by the FOSSBilling team, or a third party. 

IF, and it is a big if, we intend to create and support more payment gateways as core, then I would suggest that they should be one that covers Africa well, which according to @tobsowo would probably be Paystack, and one that covers Latin America well which would probably be Mercado Pago.

**Comment by John-S4** [comment]
Added Payfast and closed the corresponding issue

**Comment by tobsowo** [comment]
> Nice job compiling the list @BelleNottelling - I think it's a lot more useful to have them all here than to have multiple feature requests scattered around. 
> 
> A few thoughts...
> 
> It might be useful to also add a list of the currently supported payment gateways and if they are maintained by the FOSSBilling team, or a third party. 
> 
> IF, and it is a big if, we intend to create and support more payment gateways as core, then I would suggest that they should be one that covers Africa well, which according to @tobsowo would probably be Paystack, and one that covers Latin America well which would probably be Mercado Pago.

That works.
Thank you.

**Comment by roger-net** [comment]
Mercado Pago is used a lot not only here in Brazil but throughout Latin America and would be very useful for us. Below is the link to the documentation in English. Thanks.

https://www.mercadopago.com.br/developers/en/docs

**Comment by evrifaessa** [comment]
[Hyperswitch](https://hyperswitch.io/) seems like the perfect choice. It's open source, acts as a middleman to 100+ payment processors and seems easy to implement. It can also be self-hosted.

**Comment by g00g1** [comment]
Fondy ([documentation](https://portal.fondy.eu/mportal/#/docs)) and [Liqpay](https://liqpay.ua/) ([documentation](https://www.liqpay.ua/en/documentation/api/aquiring)) are pretty popular in Ukraine.

**Comment by lamanrasmi** [comment]
May I know is there any so called 'template' files that I can use to integrate my local one-off payment gateway? Most of them are using credentials matching such as: secret-key, collection-id, signature-key. And use normal callback to send back the payment status code. I do have a few sandbox/dev acc for some of the providers here in Malaysia, such as billplz, toyyibpay, paydirect, etc.

**Comment by BelleNottelling** [comment]
> May I know is there any so called 'template' files that I can use to integrate my local one-off payment gateway? Most of them are using credentials matching such as: secret-key, collection-id, signature-key. And use normal callback to send back the payment status code. I do have a few sandbox/dev acc for some of the providers here in Malaysia, such as billplz, toyyibpay, paydirect, etc.

We don't have a 'template' like that, but you can use existing payment Gateways as reference.
I'd recommend looking at the Mollie gateway: https://github.com/FOSSBilling/Mollie
Or some of the other ones (though these are mostly older code)
https://github.com/FOSSBilling/FOSSBilling/tree/main/src/library/Payment/Adapter

**Comment by binsaryunus** [comment]
Popular merchant payment gateway base on indonesia market Duitku ([doc](https://docs.duitku.com/)), Trippay ([doc](https://tripay.co.id/developer)), and ipaymu ([doc](https://documenter.getpostman.com/view/7508947/SWLfanD1?version=latest))
i have active merchant duitku that greatfull to help integrate to fossbilling

**Comment by twixi** [comment]
fondy,liqpay

**Comment by t3mp30r3k** [comment]
> Popular merchant payment gateway base on indonesia market Duitku ([doc](https://docs.duitku.com/)), Trippay ([doc](https://tripay.co.id/developer)), and ipaymu ([doc](https://documenter.getpostman.com/view/7508947/SWLfanD1?version=latest)) i have active merchant duitku that greatfull to help integrate to fossbilling

hi @binsaryunus please let me know how to integrated... is there manual or docs ?

**Comment by binsaryunus** [comment]
> > Popular merchant payment gateway base on indonesia market Duitku ([doc](https://docs.duitku.com/)), Trippay ([doc](https://tripay.co.id/developer)), and ipaymu ([doc](https://documenter.getpostman.com/view/7508947/SWLfanD1?version=latest)) i have active merchant duitku that greatfull to help integrate to fossbilling
> 
> hi @binsaryunus please let me know how to integrated... is there manual or docs ?
> 

Docs if you interested i will help it i already build duitku payment but I stuck some issue

**Comment by t3mp30r3k** [comment]
> > > Popular merchant payment gateway base on indonesia market Duitku ([doc](https://docs.duitku.com/)), Trippay ([doc](https://tripay.co.id/developer)), and ipaymu ([doc](https://documenter.getpostman.com/view/7508947/SWLfanD1?version=latest)) i have active merchant duitku that greatfull to help integrate to fossbilling
> > 
> > 
> > hi @binsaryunus please let me know how to integrated... is there manual or docs ?
> 
> Docs if you interested i will help it i already build duitku payment but I stuck some issue

what issue ?
this is a chat message from duitku, they said its not compatible.. :(
![image](https://github.com/FOSSBilling/FOSSBilling/assets/17773909/3f8cde74-4293-4ad9-8e6b-3a15c106cbd2)

**Comment by binsaryunus** [comment]
> > > > Popular merchant payment gateway base on indonesia market Duitku ([doc](https://docs.duitku.com/)), Trippay ([doc](https://tripay.co.id/developer)), and ipaymu ([doc](https://documenter.getpostman.com/view/7508947/SWLfanD1?version=latest)) i have active merchant duitku that greatfull to help integrate to fossbilling
> > > 
> > > 
> > > hi @binsaryunus please let me know how to integrated... is there manual or docs ?
> > 
> > Docs if you interested i will help it i already build duitku payment but I stuck some issue
> 
> what issue ?
> this is a chat message from duitku, they said its not compatible.. :(
> ![image](https://github.com/FOSSBilling/FOSSBilling/assets/17773909/3f8cde74-4293-4ad9-8e6b-3a15c106cbd2)
> 

Yes because fossbilling under developer/ beta so duitku don't notice it, you can build plugin yourself hire a developer.

**Comment by binsaryunus** [comment]
> > > > Popular merchant payment gateway base on indonesia market Duitku ([doc](https://docs.duitku.com/)), Trippay ([doc](https://tripay.co.id/developer)), and ipaymu ([doc](https://documenter.getpostman.com/view/7508947/SWLfanD1?version=latest)) i have active merchant duitku that greatfull to help integrate to fossbilling
> > > 
> > > 
> > > hi @binsaryunus please let me know how to integrated... is there manual or docs ?
> > 
> > Docs if you interested i will help it i already build duitku payment but I stuck some issue
> 
> what issue ?
> this is a chat message from duitku, they said its not compatible.. :(
> ![image](https://github.com/FOSSBilling/FOSSBilling/assets/17773909/3f8cde74-4293-4ad9-8e6b-3a15c106cbd2)
> 

My code need lock db from response duitku

**Comment by BRGustavoRibeiro** [comment]
I would like to suggest some payment providers focused on the Brazilian market.

- [PagSeguro](https://pagseguro.uol.com.br/)
Wonderful docs, really easy to implement. I believe that it is currently the biggest payment provider on the country.

- [Mercado Pago](https://www.mercadopago.com.br/developers/pt)
Great docs too, easy to implement.

- [Pagar.me](https://docs.pagar.me/)
Amazing platform with good fees. This one is a must for me. Docs are kinda shitty, but it is also easy (and worth it) to implement.

**Comment by Successfulsebunya** [comment]
Flutterwave is very important for Uganda and many more african countries. If only we could get a direct Mobile money intergration option. This is best for Uganda

**Comment by shmshd** [comment]
@albinvar could help us with Razorpay as he has built a working payment gateway for BoxBilling.

**Comment by BelleNottelling** [comment]
> @albinvar could help us with Razorpay as he has built a working payment gateway for BoxBilling.

It should be straightforward to port the BoxBilling one to FOSSBilling & get it on the extension directory.

**Comment by albinvar** [comment]
> > @albinvar could help us with Razorpay as he has built a working payment gateway for BoxBilling.
> 
> It should be straightforward to port the BoxBilling one to FOSSBilling & get it on the extension directory.

Should I go ahead and create the extension under my account, or will Fossbilling create a repo for Razorpay?

**Comment by evrifaessa** [comment]
> Should I go ahead and create the extension under my account, or will Fossbilling create a repo for Razorpay?

@albinvar - Hey, please create it under your GitHub account. We would be more than willing to have it listed in the extension store. Have a nice day. 🙂

**Comment by albinvar** [comment]
Thanks, @evrifaessa! I made a few adjustments, and now it's compatible with FOSSBilling. You can find the payment gateway as a separate resource on https://github.com/albinvar/Razorpay-FOSSBilling.

 I've also requested to include it in the [extension directory](https://github.com/FOSSBilling/extension-directory/pull/119).

It's worth noting that the payment gateway might be impacted by the same issue discussed in https://github.com/FOSSBilling/FOSSBilling/issues/1630.

**Comment by Anuril** [comment]
> It's worth noting that the payment gateway might be impacted by the same issue discussed in #1630.

Are you sure? Did you copy any code from the Stripe Gateway?

**Comment by albinvar** [comment]
> Are you sure? Did you copy any code from the Stripe Gateway?

Thats true, The [Razorpay module](https://github.com/albinvar/Razorpay-FOSSBilling) was created by referencing the existing [Stripe module](https://github.com/boxbilling/boxbilling/blob/master/src/bb-library/Payment/Adapter/Stripe.php) in Boxbilling, resulting in very similar code.

Moreover, I believe the issue may originate from the framework itself. Upon reviewing the process, it's clear that during the client's checkout, the sessions are either absent or cleared out when processing the payment.

What's particularly unusual is that the payment works fine when an admin is logged in. This hints that something which was once accessible to clients is now restricted, which is now requiring an admin session.

I could be mistaken in my assumptions, and I don't have much experience in dealing with such issues, but I'm willing to give it a try to resolve this problem. ✌️

**Comment by Anuril** [comment]
This is more information that we had before - and could explain a lot. 

I'm also looking into it but please continue the good work :)

**Comment by albinvar** [comment]
> > Are you sure? Did you copy any code from the Stripe Gateway?
> 
> Thats true, The [Razorpay module](https://github.com/albinvar/Razorpay-FOSSBilling) was created by referencing the existing [Stripe module](https://github.com/boxbilling/boxbilling/blob/master/src/bb-library/Payment/Adapter/Stripe.php) in Boxbilling, resulting in very similar code.
> 
> Moreover, I believe the issue may originate from the framework itself. Upon reviewing the process, it's clear that during the client's checkout, the sessions are either absent or cleared out when processing the payment.
> 
> What's particularly unusual is that the payment works fine when an admin is logged in. This hints that something which was once accessible to clients is now restricted, which is now requiring an admin session.
> 
> I could be mistaken in my assumptions, and I don't have much experience in dealing with such issues, but I'm willing to give it a try to resolve this problem. ✌️

The issue appears to be resolved for [Razorpay module](https://github.com/albinvar/Razorpay-FOSSBilling)  as well (https://github.com/FOSSBilling/FOSSBilling/pull/1650)

**Comment by t4ba** [comment]
Please add https://www.payhere.lk for Sri Lanka

**Comment by abusayed0206** [comment]
Bkash for Bangladesh. 
Documentation: [Documentation](https://developer.bka.sh/)

**Comment by sebanieto27** [comment]
MercadoPago is the most used gateway in all of Latin America. I think it would be a very important option to add and its documentation allows it to be added easily. 
https://www.mercadopago.com.br/developers/en/docs

**Comment by warmaster2143** [comment]
Is there any update for when the Payfast Payment Gateway will be added

**Comment by jaapmarcus** [comment]
> Is there any update for when the Payfast Payment Gateway will be added

Feel free to add it your self there is no plans for a developer  maintained version at the moment... We can't maintain 100 different versions..

**Comment by GreenArrowlol** [comment]
PayU (Great alternative of razorpay to those who weren't accepted by razorpay)
https://docs.payu.in

**Comment by OfficalMINI** [comment]
Upvote for PayFast

**Comment by hostserverng** [comment]
The addition of Cryptomus or Plisio cryptocurrency gateways would be awesome as well.

**Comment by araaaabbb** [comment]
I would love if you could add Xendit Gateway. These are popular payment gateways in South East Asia.
The API references are listed here [https://developers.xendit.co/api-reference/#payments-api](https://developers.xendit.co/api-reference/#payments-api)

**Comment by Palymer** [comment]
maybe FreeKassa module?

**Comment by evrifaessa** [comment]
> Please Support [Mollie](https://www.mollie.com/)

@L3xit we already support Mollie: https://github.com/fossbilling/mollie

**Comment by Raptor3um** [comment]
BTCPay would be awesome to see!

**Comment by ChristianGabs** [comment]
> BTCPay would be awesome to see!

https://extensions.fossbilling.org/

Already exist

**Comment by Palymer** [comment]
https://github.com/Palymer/FossBillingFKPayment 
made a payment module for FreeKassa, not sure if it works completely correctly, but it works

**Comment by Successfulsebunya** [comment]
Hello team. Thank you for the countless efforts. Any progress with "Flutterwave"? I still can't find it.

**Comment by jaapmarcus** [comment]
We don't actively develop modules for it...

**Comment by Successfulsebunya** [comment]
As a non developer is there a way you can guide me to make one. so the system can be fully usable in Africa and East africa mainly

**Comment by mitexleo** [comment]
Kindly add support for Paddle. It's an stripe alternative. 

Web: https://www.paddle.com/

**Comment by Abe-Telo** [comment]
What about authorize.net gateway?

**Comment by evrifaessa** [comment]
A few to add:
* PAYCOMET (#2612)
* Stancer (#2610)
* Payplug (#2613)
* Alipay (#2611)

**Comment by utkarshgual7** [comment]
Hi, I’d be happy to help with **integrating Razorpay** into the project. Let me know how I can get started or if there are any guidelines to follow.

**Comment by BelleNottelling** [comment]
> Hi, I’d be happy to help with **integrating Razorpay** into the project. Let me know how I can get started or if there are any guidelines to follow.

There was already one made.
https://github.com/albinvar/Razorpay-FOSSBilling

It might need minor updates, I am not sure

**Comment by roger-net** [comment]
With the help of AI, I created a payment module for Mercado Pago. The module is directing to the Mercado Pago portal correctly, but after the payment is made, the FossBilling status does not change. Could any programmer help resolve this? If anyone can help, I will share the module.

Com a ajuda da IA, eu criei um módulo de pagamento para o Mercado Pago, o módulo está direcionando para o portal do Mercado Pago corretamente, mas depois do pagamento efetuado, o status do FossBilling não é alterado. Algum programador poderia ajudar a resolver? Se alguém puder ajudar, eu compartilho o módulo.

**Comment by BelleNottelling** [comment]
Hi @roger-net 

late comment here.. but I think you were trying to do so around the time there was a bug in FOSSBilling that could effect it's ability to mark the invoice as paid. Might want to give it a try now, to see if things have changed

**Comment by davisvawiz** [comment]
Hi everyone,

Following the earlier interest around Paystack support for African markets, I’ve implemented a working Paystack payment gateway adapter for FOSSBilling.

Current functionality:

- One-time invoice payments
- Paystack popup checkout integration
- Automatic payment verification
- Automatic invoice processing
- Redirect back to paid invoice page
- Amount and currency validation
- Duplicate transaction protection

The current implementation has been tested and is functioning correctly for invoice payment workflows.

Project repository:

 [FOSSBilling Paystack Gateway](https://github.com/davisvawiz/FOSSBilling-Paystack-Gateway)

Contributions, suggestions, bug reports, and improvements are welcome. I’m actively maintaining and improving it, and community feedback can help strengthen the implementation further.
